### PR TITLE
Signup: Fix an error that is caused by refreshing during signup.

### LINKED ISSUE #16199: NUX: Browser refresh during signup causes broken flow
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/start
2. At any point after completing Step 1 of the process refresh the browser
3. Attempt to complete the signup

#### What I expected
To be able to finish signing up

#### What happened instead
The flow hangs on the signup processing page after inputting account information and never completes.  If you then manually navigate to wordpress.com you find that you have an account and are logged in, but in some cases (depending on where you were in the signup process when you refreshed the browser) you have no sites created.

This is likely related to https://github.com/Automattic/wp-calypso/issues/10738 and https://github.com/Automattic/wp-calypso/issues/16156

Found by @alisterscott while testing https://github.com/Automattic/wp-e2e-tests/pull/625

#### Browser / OS version
Seen in OSX Chrome and Firefox latest

#### Context / Source
#manual-testing

### PR DESCRIPTION
**Pull request by taggon**
Resolve #16199

If you refresh the browser at any point after completing step 1 of sign up process, the process never completes due to a fatal error that occurs on the confirmation page. It is because refreshing the page causes lack of some required dependency information.

This commit redirects to the first step if any mandatory data is missed.

See #16199 for more detail.

## How to test

1. Starting at URL: https://wordpress.com/start
2. At any point after completing Step 1 of the process refresh the browser
3. ~~The page should be redirected to the first step.~~  
  **UPDATED!** You should be able to resume the process where you left off.
4. Attempt to complete the signup

### CODE REVIEW

**Comment by fditrapani** [comment]
cc @designsimply @scruffian @alisterscott

**Comment by fditrapani** [comment]
Works as described 👍

**Comment by alisterscott** [comment]
Thanks for the fix. Works as described but still redirects to the first step which is less convenient than just showing the current step on page refresh. Is there anything we can do to persist the user's spot in the flow?

signup

**Comment by taggon** [comment]
This PR is updated as per @alisterscott's suggestion. Now it restores the choices you made when a signup page is refreshed so you can resume the process where you left off.

File: client/lib/signup/flow-controller.js [hunk 1]
```diff
@@ -221,6 +226,20 @@ assign( SignupFlowController.prototype, {
 		return this._flow.destination;
 	},
 
+	_getStoredDependencies() {
+		const requiredDependencies = this._flow.steps.reduce( ( current, stepName ) => {
+			if ( ! steps[ stepName ] || ! steps[ stepName ].providesDependencies ) {
```
Comment (omarjackman) [inline client/lib/signup/flow-controller.js]: lodash `get` might make this look easier on the eyes. `get( steps, [ stepName, 'providesDependencies' ] )`

```
const providesDependencies = get( steps, [ stepName, 'providesDependencies' ] );
return providesDependencies ? current.concat( providesDependencies ) : current;
```

File: client/lib/signup/progress-store.js [hunk 2]
```diff
@@ -132,6 +132,24 @@ function handleChange() {
...
+	let unstorableDependencies;
+
+	if ( steps[ step.stepName ] ) {
+		unstorableDependencies = steps[ step.stepName ].unstorableDependencies;
+	}
+
+	if ( isEmpty( action.providedDependencies ) ) {
+		return step;
+	}
+
```
Comment (omarjackman) [inline client/lib/signup/progress-store.js]: You could get rid of the `if` and `let` and just do
```
const unstorableDependencies = get( steps, [ step.stepName, 'unstorableDependencies' ] );
```

**Comment by alisterscott** [comment]
Thanks for the fix with the step where you're up to: tested and works well

**Comment by taggon** [comment]
@omarjackman Thanks! Updated. :)

**Pull request merged**